lib/Sema/CSDiagnostics.cpp

```diff
@@ -110,6 +110,161 @@ Optional<SelectedOverload> FailureDiagnostic::getChoiceFor(Expr *expr) const {
   return getOverloadChoiceIfAvailable(cs.getCalleeLocator(expr));
 }
 
+Type FailureDiagnostic::resolveInterfaceType(Type type,
+                                             bool reconstituteSugar) const {
+  auto &cs = getConstraintSystem();
+  auto resolvedType = type.transform([&](Type type) -> Type {
+    if (auto *tvt = type->getAs<TypeVariableType>()) {
+      auto *loc = tvt->getImpl().getLocator();
+
+      // If this type variable is for a generic parameter, return that.
+      if (loc->isForGenericParameter())
+        return loc->getGenericParameter();
+
+      // Otherwise resolve its fixed type, mapped out of context.
+      if (auto fixed = cs.getFixedType(tvt))
+        return resolveInterfaceType(fixed->mapTypeOutOfContext());
```

Comment:
I think it might make more sense to do `auto type = resolveType(tvt); return resolveInterfaceType(type->mapOutOfContext());` here because type variable could be bound to array where element was another type variable or something...

Reply:
@xedin I might be mistaken, but couldn't that cause us to "miss" a type variable that opens a generic parameter? For example if we have `$T0` bound to `[$T1]`, where `$T1` opens a generic placeholder `T`, and could be bound to say `Int`. If we call `resolveType` on `$T0` we'd get back `[Int]` rather than `[T]`.

That being said, I'm can't seem to find an example that does this, so it might not be a concern.

Reply:
Sorry yes, I misread this one, you want it to recurse and find my type variables! One more suggestion - you can simplify checking for generic parameter into `if (auto *GP = tvt->getImpl().getGenericParameter()) return GP;`

Reply:
@xedin Ah nice! Thanks, I've gone ahead and made that change.